### PR TITLE
Feature/charging with adyen

### PR DESCRIPTION
**Pull request by meletios**

### CODE REVIEW

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+
+    } catch (ex: Exception) {
+      commandGateway.sendAndWait<Void>(
+        CancelAdyenTransactionCommand(
+          e.transactionId,
+          e.memberId,
+          e.recurringDetailReference,
+          e.amount,
+          EXCEPTION_MESSAGE
```
Comment (FredrikAreschoug) [inline payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt:97]: can and should we maybe include the message from `ex: Exception` or at least something in the reason?

File: bin/docker_push.sh [hunk 2]
```diff
@@ -2,7 +2,7 @@
 # Push only if it's not a pull request
 if [ -z "$TRAVIS_PULL_REQUEST" ] || [ "$TRAVIS_PULL_REQUEST" == "false" ]; then
   # Push only if we're testing the master branch
-  if [ "$TRAVIS_BRANCH" == "master" ]; then
+  if [ "$TRAVIS_BRANCH" == "master" ] || [ "$TRAVIS_BRANCH" == "feature/kotlinize-member-aggregate" ]; then
```
[the remark this answers is not in the document]
Comment (meletios) [reply bin/docker_push.sh:5]: 😇

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt (hunk 1 above)
Comment (meletios) [reply payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt]: correct, i added it

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/payments/commands/ChargeCompletedCommand.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.hedvig.paymentservice.domain.payments.commands
+
+import lombok.Value
+import org.axonframework.commandhandling.TargetAggregateIdentifier
+import java.time.Instant
+import java.util.UUID
+import javax.money.MonetaryAmount
+
+@Value
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/domain/payments/commands/ChargeCompletedCommand.kt:9]: 👀

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/payments/commands/CreatePayoutCommand.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.hedvig.paymentservice.domain.payments.TransactionCategory
+import lombok.Value
+import org.axonframework.commandhandling.TargetAggregateIdentifier
+import java.time.Instant
+import java.time.LocalDate
+import java.util.UUID
+import javax.money.MonetaryAmount
+
+@Value
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/domain/payments/commands/CreatePayoutCommand.kt:11]: 👀

File: payment-service/src/main/java/com/hedvig/paymentservice/serviceIntergration/productPricing/dto/MarketInfo.kt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.hedvig.paymentservice.serviceIntergration.productPricing.dto
+
+import javax.money.CurrencyUnit
+
+data class MarketInfo(
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/serviceIntergration/productPricing/dto/MarketInfo.kt:5]: We do not want to confuse market as a set in stone concept, this one is related to the contract

```suggestion
data class ContractMarketInfo(
```

File: payment-service/src/main/java/com/hedvig/paymentservice/web/internal/MemberController.java [hunk 6]
```diff
@@ -27,15 +34,24 @@
...
     this.memberRepository = memberRepository;
+    this.productPricingService = productPricingService;
   }
 
   @PostMapping(path = "{memberId}/charge")
   public ResponseEntity<?> chargeMember(
-      @PathVariable String memberId, @RequestBody ChargeRequest request) {
+    @PathVariable String memberId, @RequestBody ChargeRequest request) {
+
+    MarketInfo marketInfo = productPricingService.getMarketInfo(memberId);
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/web/internal/MemberController.java:49]:
```suggestion
    MarketInfo contractCarketInfo = productPricingService.getContractMarketInfo(memberId);
```

File: payment-service/src/main/java/com/hedvig/paymentservice/web/v2/MemberControllerV2.kt [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  @PostMapping(path = ["{memberId}/payout"])
+  fun payoutMember(
+    @PathVariable memberId: String,
+    @RequestParam(
+      name = "category",
+      required = false,
+      defaultValue = "CLAIM"
+    ) category: TransactionCategory,
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/web/v2/MemberControllerV2.kt:58]: Probably some of my first lines of code written at Hedvig 😍

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  @Transient
+  private lateinit var adyenService: AdyenService
+
+  @StartSaga
+  @SagaEventHandler(associationProperty = TRANSACTION_ID)
+  @EndSaga
+  fun on(e: AdyenTransactionInitiatedEvent) {
+    try {
+      val request = ChargeMemberWithTokenRequest(e.transactionId, e.memberId, e.recurringDetailReference, e.amount)
+      val response = adyenService.chargeMemberWithToken(request)
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt:36]: Even though I know where you come from with this and it is completely fine. I would suggest considering to instead inject `adyenService` together with the command and perform this inside the aggregate and apply an event accordingly.

Comment (vonElfvin) [reply payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt:36]: Some commands will have to be in a saga, for example those who speak to another aggregate, the other stuff should be owned by aggregate IMO

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/payments/sagas/ChargeSaga.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  @Transient
+  lateinit var trustlyService: TrustlyService
+  @Autowired
+  @Transient
+  lateinit var uuidGenerator: UUIDGenerator
+
+  @StartSaga
+  @SagaEventHandler(associationProperty = "memberId")
+  @EndSaga
+  fun on(e: ChargeCreatedEvent) {
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/domain/payments/sagas/ChargeSaga.kt:32]: Nice solution

File: payment-service/src/main/java/com/hedvig/paymentservice/web/internal/MemberController.java [hunk 10]
```diff
@@ -27,15 +34,24 @@
...
+    if (marketInfo.getPreferredCurrency() != request.getAmount().getCurrency()) {
```
Comment (vonElfvin) [inline payment-service/src/main/java/com/hedvig/paymentservice/web/internal/MemberController.java:50]: Similarly you should probably have this logic inside the aggregate. Inject productPricing into the commandhandler of starting a charge and make this validation there. This way noone can accidently charge the wrong currency from another part of the system. We can always throw as much exceptions we want inside a command handler. Doing it in the controller feels like the unsafe and wrong way from a event sourcing perspective. Return a result from the command handelr if you want to send an appropriate response in the controoller.

Comment (meletios) [reply payment-service/src/main/java/com/hedvig/paymentservice/web/internal/MemberController.java:50]: good point i will change it

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt (hunk 8 above)
Comment (meletios) [reply payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt:36]: I can see the point with adding that in the aggregate, but I will keep that to the Saga for now. 🙏

File: payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt [hunk 11]
```diff
@@ -97,7 +100,14 @@ class AdyenTransactionSaga {
...
 
-/* Comment until ADYEN will let us know when the charge will be considered final
+
+  /*
+  * When the capture delay is set to immediate and the authorisation is successful, the funds are immediately captured.
+  * Because it is immediate, Adyen only send an authorisation notification.
+  * In that scenario, we can indeed consider the transaction is successful after the authorisation.
+  *
+  * If for some reason the capture fails, we will receive a capture_failed notification as described in Adyen's documentation.
+  * */
```
Comment (FredrikAreschoug) [inline payment-service/src/main/java/com/hedvig/paymentservice/domain/adyenTransaction/sagas/AdyenTransactionSaga.kt:110]: :heart_eyes:

**Pull request merged**